### PR TITLE
Pytorch Lightning Callback

### PR DESCRIPTION
**Pull request by gan3sh500**
Created a Pytorch Lightning Callback to address issue #367 . I used a Callback instead of Logger as ModelCheckpoint is an existing callback in Pytorch Lightning. Using this is simple. 
```python
replicate_callback = ReplicateCallback({
    'dataset': 'mnist',
    'precision': 16,
    'encoder_dim': 3
}, monitor='val_loss')
trainer = pl.Trainer(gpus=1, num_nodes=1, precision=16,
                     limit_train_batches=0.5,
                     callbacks=[replicate_callback])
```
This doesn't have any checks about `monitor` and doesn't inherit from [PL's ModelCheckpoint](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/model_checkpoint.py) like you have done for Keras Integration. Please comment and let me know what would be a good way to progress in the PR.

### CODE REVIEW

**Comment by turian** [comment]
@gan3sh500 "This doesn't have any checks about monitor" Can you explain what you mean?

What are the limitations of the current approach?

**Comment by gan3sh500** [comment]
In PL they check to ensure than monitor is collected. Just to catch things like monitor = 'val_lossss'.  They also decide mode automatically, adds support for some older monitoring and allow for custom save functions. I'm not sure if these are needed or just nice to have.

File: python/replicate/pl_callback.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        self.params = params
+        self.monitor = monitor
+        self.period = period
+        self.save_weights_only = save_weights_only
+        self.mode = 'minimize' if mode == 'min' else 'maximize'
+
+    def on_pretrain_routine_start(self, trainer, pl_module):
+        self.experiment = replicate.init(path='.', params=self.params)
+
+    def on_validation_end(self, trainer, pl_module):
```
Comment (andreasjansson) [inline python/replicate/pl_callback.py:27]: I'm not a PL user myself so this is maybe a stupid question, but does PL always do validation? Or are there uses of PL where the Replicate callback wouldn't actually call `self._save_model`?

**Review by andreasjansson — APPROVED** [review]
This looks good, just a question about how PL handles training if there is no validation.

It needs a test and a bit of documentation as well, but we can add that in a second PR.

**Comment by andreasjansson** [comment]
Looks like there's some linting failures as well, if there aren't any mypy stubs for PL you can append `# type: ignore` to the import.

File: python/replicate/pl_callback.py (hunk 1 above)
Comment (turian) [reply python/replicate/pl_callback.py:27]: @andreasjansson There are definitely cases where PL doesn't do validation. e.g. you just write train code, not validation code.

For the second question, I am not sure.

Comment (turian) [reply python/replicate/pl_callback.py:27]: p.s. you should consider trying PL, it's a great framework.

Comment (gan3sh500) [reply python/replicate/pl_callback.py:27]: @andreasjansson Yes. I think I should add checks for that. I need to investigate PL's ModelCheckpoint in more scenarios. I am not personally a PL user sadly, I am just familiar with the repo from contributing. 
@turian if you have any suggestions on what else I should check for please let me know.

**Comment by gan3sh500** [comment]
@andreasjansson How do I get the linting settings used for the repo into my env? Maybe a noob question but I always find this confusing for any new repo.

**Comment by andreasjansson** [comment]
@gan3sh500 It depends on which editor you're using. If you're using VSCode, we have setup instructions in [CONTRIBUTING.md](https://github.com/replicate/replicate/blob/main/CONTRIBUTING.md#vscode-development-environment) (which is a somewhat standard place these sort of thing go).

You can also manually run the linters locally inside the `python` directory with `make lint` (which then invokes `mypy .` and `black --check .`).

**Comment by gan3sh500** [comment]
@andreasjansson I'll fix up the lint issues and finish this up and work on the testing and doc for this.

**Comment by andreasjansson** [comment]
@gan3sh500 I added some tests and made the callback be called `on_epoch_end` if no validation is defined.

@turian Would you mind taking a look since I'm a PL novice?

**Comment by gan3sh500** [comment]
@andreasjansson If there is nothing else I can add here, shall I open another PR for the Doc for this?

**Comment by andreasjansson** [comment]
Thanks @gan3sh500 yes a docs PR would be brilliant!

**Comment by andreasjansson** [comment]
Actually @gan3sh500 let's hold off on that for a bit. I think I'll rewrite this as a PL Logger instead, similar to the MLFlow and Comet loggers etc.

**Comment by andreasjansson** [comment]
For the record, we've decided to stick with the callback for now, but have plans to integrate into PL as a logger eventually: https://github.com/replicate/replicate/issues/432

File: web/pages/docs/guides/pytorch-lightning-integration.mdx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import torch
+from torch.nn import functional as F
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning import Trainer
+from torch.optim import Adam
+from torch.utils.data import DataLoader, random_split, Subset
+from torchvision.datasets import MNIST
+from torchvision import transforms
+
+from replicate.pl_callback import ReplicateCallback
```
Comment (gan3sh500) [inline web/pages/docs/guides/pytorch-lightning-integration.mdx:27]: Shall we add #highlight-line and #highlight-start and end similar to the Keras guide. It helps show how little extra code Replicate adds.

**Review by gan3sh500 — COMMENTED** [review]
Really nice that you used the PL module for everything, so many PL examples don't use the module for data. Wouldn't the spacing and a main and if __name___ etc add unnecessary length? This will be viewed from a browser not an IDE right. The style in Keras integration example helps a lot with readability.

**Comment by gan3sh500** [comment]
@andreasjansson I can't see the changes if I run `yarn; yarn build; yarn start` for your commit or mine. I'm not sure what is going on.

**Pull request merged**